### PR TITLE
Masking fix

### PR DESCRIPTION
**Pull request by LoganWright**
absorbing #13

### CODE REVIEW

File: Sources/WebSocket/WebSocket.swift [hunk 1]
```diff
@@ -6,13 +8,49 @@
...
     /// See `BasicWorker`.
     public var eventLoop: EventLoop {
         return channel.eventLoop
     }
 
     /// Outbound `WebSocketEventHandler`.
     private let channel: Channel
+    
+    /// `WebSocket` processing mode.
+    public private(set) var mode: Mode
```
Comment (tanner0101) [inline Sources/WebSocket/WebSocket.swift]: Can a websocket ever change from client to server or vice versa? Seems like this can be `let`.

Comment (tanner0101) [reply Sources/WebSocket/WebSocket.swift]: Could this be private?

File: Sources/WebSocket/WebSocket.swift [hunk 2]
```diff
@@ -6,13 +8,49 @@
...
+        /// sends to the server.
+        /// The server MUST close the connection upon receiving a frame that is not masked.
+        /// A server MUST NOT mask any frames that it sends to the client.
+        /// A client MUST close a connection if it detects a masked frame.
+        ///
+        /// RFC 6455 Section 5.3
+        /// The masking key is a 32-bit value chosen at random by the client.
+        /// When preparing a masked frame, the client MUST pick a fresh masking
+        /// key from the set of allowed 32-bit values.
+        public func makeMaskKey() -> WebSocketMaskingKey? {
```
Comment (tanner0101) [inline Sources/WebSocket/WebSocket.swift]: Can this be internal?

File: Sources/WebSocket/WebSocket.swift [hunk 3]
```diff
@@ -6,13 +8,49 @@
 ///      }
 ///
 public final class WebSocket: BasicWorker {
+    
+    /// Available WebSocket modes. Either `Client` or `Server`.
+    public enum Mode {
```
Comment (tanner0101) [inline Sources/WebSocket/WebSocket.swift]: This type doesn't seemed to be exposed anywhere. Could it be left internal?

File: Sources/WebSocket/WebSocket.swift [hunk 4]
```diff
@@ -191,7 +230,8 @@ public final class WebSocket: BasicWorker {
         let data = data.convertToData()
         var buffer = channel.allocator.buffer(capacity: data.count)
         buffer.write(bytes: data)
-        send(WebSocketFrame(fin: true, opcode: opcode, data: buffer), promise: promise)
+        let maskKey: WebSocketMaskingKey? = mode.makeMaskKey()
```
Comment (tanner0101) [inline Sources/WebSocket/WebSocket.swift]: Do we need to generate a mask key for each send? That adds quite a bit of overhead to each call. I wonder if NIO has any way of doing this for us automatically, or if we can use a static key per websocket connection.

Comment (LoganWright) [reply Sources/WebSocket/WebSocket.swift]: Yes, necessary for each one to be unique and new.

**Comment by penny-coin** [comment]
Hey @LoganWright, you just merged a pull request, have a coin! 

You now have 494 coins.

**Pull request merged**